Add CryptoKit implementation for X25519/Ed25519

nmahendru opened the pull request:
#### a16a1e8ea319fea611553fd4d6354559dfd49054
<pre>
Add CryptoKit implementation for X25519/Ed25519
<a href="https://rdar.apple.com/126582690">rdar://126582690</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272786">https://bugs.webkit.org/show_bug.cgi?id=272786</a>

Reviewed by Alex Christensen.

ed25519/x25519 cryptokit implementation.
wpt tests run fine with cryptokit enabled.

Also adding sha1 as an option for ecdsa. That was missed in the last
attempt. Some refactoring in the swift shim to group extensions.

* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(EdRv.errCode):
(EdRv.signature):
(EdRv.keyBytes):
(EdKey.sign(_:key:data:)):
(EdKey.verify(_:key:signature:data:)):
(EdKey.deriveBits(_:priv:pub:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(Curve25519.signature(_:)):
(Curve25519.isValidSignature(_:data:)):
(Curve25519.sharedSecretFromKeyAgreement(_:)):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::toCKHashFunction):
(WebCore::verifyHashParameter):
(WebCore::signECDSACryptoKit):
(WebCore::verifyECDSACryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::signEd25519CryptoKit):
(WebCore::verifyEd25519CryptoKit):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCryptoKit):
(WebCore::deriveBitsCoreCrypto):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp:
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmX25519GCrypt.cpp:
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):

Canonical link: <a href="https://commits.webkit.org/277746@main">https://commits.webkit.org/277746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79fdea02d8ea8d755d6dda93684664fe74ca6a5a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51143 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44520 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25189 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39615 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49037 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25346 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20734 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22829 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6512 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44771 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53048 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23502 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19827 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46925 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24767 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42028 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45844 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25572 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6899 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24490 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -224,4 +224,52 @@ extension P521.Signing.PrivateKey {
+        }
+        return self.isValidSignature(Data.temporaryDataFromSpan(spanNoCopy: signature), for: Data.temporaryDataFromSpan(spanNoCopy: data))
+    }
+    
```
nit: extra space

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
```diff
@@ -53,17 +53,18 @@ static PAL::HashFunction toCKHashFunction(CryptoAlgorithmIdentifier hash)
     default:
         ASSERT_NOT_REACHED();
         return PAL::HashFunction::sha512();
+        break;
```
break after return is unreachable.  Could you also remove the default above, then put ASSERT_NOT_REACHED(); and return sha12 after the switch for if an invalid value hits the switch?  Removing "default" gives a compiler warning if we've missed a case.

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
```diff
@@ -27,12 +27,17 @@
-static ExceptionOr<Vector<uint8_t>> signEd25519(const Vector<uint8_t>& sk, size_t, const Vector<uint8_t>& data)
+static ExceptionOr<Vector<uint8_t>> signEd25519(const Vector<uint8_t>& sk, const Vector<uint8_t>& data)
 {
+    if (sk.size() != ed25519KeySize)
```
Can we use std::array<uint8_t, ed25519KeySize> instead of Vector?  Same in several places.

nmahendru replied:
Soon we will have to add support for Ed448 and then the size will not be fixed. So Vector is probably better ?
I could make it a variant but then all that ifelse just pollutes the code :D

achristensen07 commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -50,7 +50,6 @@ PlatformECKeyContainer CryptoKeyEC::toCKPlatformECKeyContainer(PAL::ECKey key)
     return makeUniqueRefWithoutFastMallocCheck<PAL::ECKey>(key);
 }
 #endif
-
```
Leave this newline.

nmahendru committed:
Add CryptoKit implementation for X25519/Ed25519
rdar://126582690
https://bugs.webkit.org/show_bug.cgi?id=272786

Reviewed by Alex Christensen.

ed25519/x25519 cryptokit implementation.
wpt tests run fine with cryptokit enabled.

Also adding sha1 as an option for ecdsa. That was missed in the last
attempt. Some refactoring in the swift shim to group extensions.

* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(EdRv.errCode):
(EdRv.signature):
(EdRv.keyBytes):
(EdKey.sign(_:key:data:)):
(EdKey.verify(_:key:signature:data:)):
(EdKey.deriveBits(_:priv:pub:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(Curve25519.signature(_:)):
(Curve25519.isValidSignature(_:data:)):
(Curve25519.sharedSecretFromKeyAgreement(_:)):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::toCKHashFunction):
(WebCore::verifyHashParameter):
(WebCore::signECDSACryptoKit):
(WebCore::verifyECDSACryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::signEd25519CryptoKit):
(WebCore::verifyEd25519CryptoKit):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCryptoKit):
(WebCore::deriveBitsCoreCrypto):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp:
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmX25519GCrypt.cpp:
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):

Canonical link: https://commits.webkit.org/277746@main
